### DISCUSSION TITLE
Setting property to empty array causes it to be "undefined"

### DISCUSSION THREAD

**Discussion by quasipickle** [alpinejs/alpine#1950] [1. Help] [answered]
I'm enabling SortableJS on a list of elements that is rendered out using Alpine.  I want the option to "Cancel" the sorting, thereby reverting the list to it's pre-sorted state.  Since SortableJS doesn't affect Alpine's data, just the DOM, I figured if I just emptied out Alpine's data, then re-set it, that would cause Alpine to re-render the list in the original order.

After resetting, I get an error of `Alpine Expression Error: Cannot read property 'length' of undefined...`.  However, the only thing I access `.length` on is `componentItems`, and `componentItems` is never undefined.

Why is `componentItems` ever being evaluated as undefined?

**Runnable test case: https://codepen.io/quasipickle/pen/BaREQvR?editors=1011**

Test case markup:
```html
<div x-data="sorter">
  <button type = "button" @click = "reorder()">Reorder</button>
  <button type = "button" @click = "resetComponent()">Reset</button>
  <template x-if="componentItems.length > 0">
      <div id = "list">
        <template x-for="item in componentItems" :key="item.id">
          <div x-text="item.name"></div>
        </template>
      </div>
  </template>
</div>
```
Test case javascript:
```javascript
document.addEventListener('alpine:init',()=>{
  Alpine.data("sorter",()=>({
    componentItems:[
      { id:0, name:"A" },
      { id:1, name:"B" },
      { id:2, name:"C" }
    ],
    resetComponent(){
      temp = this.componentItems;
      this.componentItems = [];
      this.$nextTick(()=>{
        this.componentItems = temp;
      });
    },
    reorder(){
      const $list = document.querySelector('#list');
      const $$divs = $list.getElementsByTagName('div');
      $list.insertBefore($$divs[0],$$divs[$$divs.length]);
    }
  }));
});
```

**Edit:** This looks related to https://github.com/alpinejs/alpine/discussions/1917

**Comment by pohlt** [comment]
Your chances for someone looking into this issue are much higher if you provide an example at [CodePen](https://codepen.io/). 😉

**Reply by quasipickle** [reply]
That's why I linked to a codepen example above the code blocks 😉 .  I'll bold it in case others miss it too.

Thanks for taking the time to try to help.

**Reply by pohlt** [reply]
D'oh! 🤦🏼

**Reply by pohlt** [reply]
Just an observation: If you remove the first `template` with `x-if`, the second `template` with `x-for` does not have any issues with the `componentItems` variable. Weird. 🤔 
If you can live with an empty list being "displayed", this might be a workaround.

**Reply by quasipickle** [reply]
I know the feeling.

Removing the `x-if` isn't an option.  While it would work in this simplified test-case, the real world implementation has quite a bit of UI inside that `x-if`.

**Reply by pohlt** [reply]
Looks like you are not the only one: #1917, #1837
Maybe it would work with V2? Not a long-term solution, I know.

**Reply by quasipickle** [reply]
Yeah, I saw those after I posted.  The problem was reported over 2 weeks ago with no response so, I dunno - start looking at Vue maybe?

**Comment by KevinBatdorf** [comment, marked as answer]
When Alpine `x-if` removes the inner content it's not preserving the keys for the `x-for`. There's probably room for some improvement there within Alpine! For now you should use `<div x-show="componentItems && componentItems.length > 0">`

**Reply by quasipickle** [reply]
Thanks for this.  It worked great.

However, there are still some questions/problems:

1. From the documentation, it seems like it shouldn't matter what the keys are.  It says the elements are completely removed from the DOM, so inner code shouldn't be evaluated at all.  But you probably already knew that from the myriad discussions about this.
2. What's also interesting is not just that the code inside `x-if` is being evaluated, but that it's saying the item is undefined.  In the code example above, at no point is `componentItems` undefined.
3. Finally, why does this work with `x-show`?  The documentation just says it shows and hides, not that it affects the DOM at all.  Looking at the source code though, I can see the DOM nodes appearing and disappearing.  `x-show` seems to be doing exactly what I would expect `x-if` to do.

**Comment by MuzafferDede** [comment]
@quasipickle I got some time to play around with this issue. Alternatively you could use `setTimeout` instead of `$nextTick` 
```js
...
resetComponent() {
    temp = this.componentItems;

    this.componentItems = [];

    setTimeout(() => {
        this.componentItems = temp;
    });
},
...
```

**Reply by quasipickle** [reply]
Thanks for this.  @KevinBatdorf 's solution worked great so I don't need to try this.  I appreciate the help nonetheless.
